core-util/v2/impl/FP0.hpp

```diff
@@ -0,0 +1,238 @@
+        reinterpret_cast<impl::FunctionPointerInterface<R()>*>(_storage)->~FunctionPointerInterface<R()>();
+        rhs.copy_to(_storage);
+        return *this;
+    }
+    inline R operator ()() const {
+        return call();
+    }
+    R call() const {
+        return reinterpret_cast<impl::FunctionPointerInterface<R()>*>(_storage)->call();
+    }
+    ~FunctionPointer() {
+        reinterpret_cast<impl::FunctionPointerInterface<R()>*>(_storage)->~FunctionPointerInterface<R()>();
+    }
+protected:
+    union {
+        char _storage[sizeof(impl::FunctionPointerSize<I>)];
+        impl::FunctionPointerSize<I> _alignementAndSizeGuarantees;
+    };
+};
+
+namespace impl {
+template <typename R, size_t I>
+class FunctionPointerInterface<R(),I>{
+public:
+    virtual operator bool() const = 0;
+    virtual bool operator ==(const FunctionPointerInterface &) const = 0;
+    virtual R call(void) const = 0;
+    virtual ~FunctionPointerInterface(){};
+    virtual void copy_to(void *) const = 0;
+protected:
+    FunctionPointerStorage<I> _fp;
+};
+
+/** A class for storing and calling a pointer to a static or member void function without arguments
+ */
+template<typename R, size_t I>
+class StaticPointer<R(), I> : public FunctionPointerInterface<R(), I>{
+public:
+    StaticPointer(R (*function)() = 0)
+    {
+        attach(function);
+    }
+    StaticPointer(const StaticPointer& fp) {
+        *this = fp;
+    }
+    void attach(R (*function)()) {
+        this->_fp._static_fp = (void*)function;
+    }
+    R call() const{
+        return reinterpret_cast<R (*)()>(this->_fp._static_fp)();
+    }
+    R (*get_function() const)() {
+        return this->_fp._static_fp;
+    }
+    operator bool() const {
+        return (this->_fp._static_fp != NULL);
+    }
+    bool operator==(const FunctionPointerInterface<R()> & rhs) const {
+        return (this->_fp._static_fp == static_cast<const StaticPointer *>(&rhs)->_fp._static_fp);
+    }
+    StaticPointer & operator = (const StaticPointer & rhs) {
+        this->_fp._static_fp = rhs._fp._static_fp;
+        return *this;
+    }
+    void copy_to(void * storage) const {
+        new(storage) StaticPointer(*this);
+    }
+};
+
+template<typename R, typename C, size_t I>
+class MethodPointer<R(), C, I> : public FunctionPointerInterface<R(), I>{
+public:
+    MethodPointer(C *object, R (C::*member)(void))
+    {
+        attach(object, member);
+    }
+    MethodPointer(const MethodPointer& fp) {
+        *this = fp;
+    }
+    void attach(C *object, R (C::*member)(void)) {
+        this->_fp._method._object = (void *) object;
+        *reinterpret_cast<R (C::**)(void)>(this->_fp._method._member) = member;
+    }
+    R call() const{
+        C* o = static_cast<C*>(this->_fp._method._object);
+        R (C::**m)(void) = reinterpret_cast<R (C::**)(void)>(const_cast<char *>(this->_fp._method._member));
+        return (o->**m)();
+    }
+    R (*get_function() const)(){
```

Comment:
Can we do without having this method for MethodPointer and others (except StaticPointer)? else if it is generic, it should be in FunctionPointerInterface